### ISSUE TITLE
Support AWS V2 Java API

### ISSUE THREAD

**Issue by notnarb** [google/tink#506]
TL;DR: AWS maintains 2 versions of their Java AWS libraries, I'm willing to submit a new client for the 2nd version.

**Help us help you**

Our team is using Tink for hybrid encryption and one of our planned deployment targets are AWS Lambda functions in Java.

**Is your feature request related to a problem?**

Currently the [AWS KMS implementation](https://github.com/google/tink/blob/master/java_src/src/main/java/com/google/crypto/tink/integration/awskms/AwsKmsClient.java) provided with Tink for Java uses the V1 API `com.amazonaws.services.*`

_Note: as far as I am aware, the "V2" Java API is a difference in client implementation and not a separate HTTP endpoint with different behavior_

The big advantages I see for Tink offering an integration with the [V2 API](https://docs.aws.amazon.com/sdk-for-java/latest/migration-guide/whats-different.html) client (`software.amazon.awssdk.*`) are:
 * Allowing replacement of the HTTP client implementation (["pluggable http layer"](https://aws.amazon.com/blogs/developer/aws-sdk-for-java-2-x-released/))
   * HTTP client can also be reused between V2 clients
 * Decreased jar size for apps already using V2 of the API
 * Credential provider can be reused with V2 clients

These features are important for startup-latency-sensitive environments like ours (Lambda functions), but I'll admin these are probably not very important for the majority of AWS x Tink x Java users.

**Describe the solution you'd like**

 * A new AwsKmsV2Client (and corresponding AwsKmsV2Aead) class specifically for supporting the V2 Java AWS client.
 * A new method `withHttpClient()` unique to this client.

cons: increased surface area, more code to maintain

**Describe alternatives you've considered**

1) Replacing the V1 client with the V2 client.

Less code to maintain and presumably the V2 client is better _but_ this would be a **breaking change** because `withCredentialsProvider` currently accepts a V1 provider.  Maybe that's fine and consumers can find a way to translate their V1 provider into a V2?

2) Updating the current AwsKmsClient to support both API types

All of the types (e.g. [AwsCredentialsProvider](https://github.com/google/tink/blob/58be99b3c4d09154d12643327f293cc45b2a6a7b/java_src/src/main/java/com/google/crypto/tink/integration/awskms/AwsKmsClient.java#L48)) are different between the two API versions so the implementation will likely become very bloated.  Also this increases the Jar size.

**Additional context**

I have a relatively simple working PoC made already that I would be willing to tidy up to contribute if my proposed solution sounds reasonable.

Also happy to chat about this if you'd like, feel free to reach out internally.

**Comment by foomeden** [comment]
+1 on this - extra benefit of AWS SDK v2 is the support for k8s ServiceAccount-aware credentials provider (enables granting granular permissions on KEKs to individual pods).

**Comment by juergw** [comment]
Thanks for reporting this. This is certainly something we will need to look at. But we also plan to do other changes to the Tink's KMS API. So this may take some time.

**Comment by notnarb** [comment]
As a quick update my implementation using v2 of the SDK is now open source and can be found at https://github.com/privacysandbox/control-plane-shared-libraries/tree/main/java/com/google/crypto/tink/integration/awskmsv2 ([permalink](https://github.com/privacysandbox/control-plane-shared-libraries/blob/e93a15281714ed7da45a80a1d2a1e8c9849aa929/java/com/google/crypto/tink/integration/awskmsv2/AwsKmsV2Aead.java#L40))

> But we also plan to do other changes to the Tink's KMS API. So this may take some time.

:+1: I would love to see this code upstream to get more eyes on it and remove the need for us to maintain it but we are unblocked right now.

Since this code is Apache licensed, others can probably unblock themselves by referring to this implementation.

---

Also included (but maybe doesn't make sense to upstream?) are simple API-backed implementations of PublicKeySign and PublicKeyVerify

**Comment by juergw** [comment]
Great, thanks for letting us know!
